Fix AddressController.php

metacreo opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           |  8.1.x
| Description?      | Duplicate function FillWith call in AddressController.php. Address form filled two times. First time normal, second time with empty country. If default country is US, form shows needed states to enter. Fix bugs: https://github.com/PrestaShop/PrestaShop/issues/35013, https://github.com/PrestaShop/PrestaShop/issues/34080, https://github.com/PrestaShop/PrestaShop/issues/33991
| Type?             | bug fix 
| Category?         | FO 
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Indicate how to verify that this change works as expected.
| UI Tests          | Please run UI tests and paste here the link to the run. [Read this page to know why and how to use this tool.](https://devdocs.prestashop-project.org/8/contribute/contribution-guidelines/ui-tests/).
| Fixed issue or discussion?     | Fixes #{issue number here}, Fixes #{another issue number here}, Fixes https://github.com/PrestaShop/PrestaShop/discussions/ {#35013 #34080 #33991}

metacreo committed:
Fix AddressController.php

Duplicate call function. Address form filled second time. Fix bugs: 
https://github.com/PrestaShop/PrestaShop/issues/34080
and
https://github.com/PrestaShop/PrestaShop/issues/33991

mflasquin reviewed (commented):
Seems good, but can you tell us where the address form is already filled in the code ? thanks

metacreo commented:
@mflasquin 
AddressController.php
```
   public function postProcess()
    {
        $this->context->smarty->assign('editing', false);
        $id_address = (int) Tools::getValue('id_address');
        // Initialize address if an id exists
        if ($id_address) {
            $this->address_form->loadAddressById($id_address);
        }

        // Fill the form with data
        $this->address_form->fillWith(Tools::getAllValues());
```
CustomerAddressForm.php
```
   public function loadAddressById($id_address)
    {
        $context = Context::getContext();

        $this->address = new Address($id_address, $this->language->id);

        if ($this->address->id === null) {
            return Tools::redirect('pagenotfound');
        }

        if (!$context->customer->isLogged() && !$context->customer->isGuest()) {
            return Tools::redirect('/index.php?controller=authentication');
        }

        if ($this->address->id_customer != $context->customer->id) {
            return Tools::redirect('pagenotfound');
        }

        $params = get_object_vars($this->address);
        $params['id_address'] = $this->address->id;

        return $this->fillWith($params);
    }

    public function fillWith(array $params = [])
    {
```

mflasquin reviewed (dismissed):
LGTM

kpodemski commented:
Hello @metacreo 

It looks legit. What I suggest now:
- rebase into 8.1.x branch, so that we can all have this fix before PrestaShop 9 :)
- help QA reproduce this issue without using demo.prestashop.com, the origin of the installation shouldn't matter - so kindly provide them with the steps that allow them to reproduce the issue - from what I see the confusion is between adding and editing the address

metacreo commented:
rebase ok, but I don`t known how to help QA. They cannot reproduce, although I reproduced on 3 production servers.

Hlavtox commented:
Hi @metacreo, if it happens only in some browsers, I don't know how JS error is related to removing something on PHP side?

metacreo commented:
@Hlavtox Hello. 
The fact of the matter is that, for example, on server 1 I only have it with Firefox, on server 2 only with Chrome, and on server 3 on both browsers. PHP versions are identical, PS versions identical. I think that this difference is still related to PHP.

matthieu-rolland reviewed (commented):
thank you for your contribution @metacreo 👍 

Would you mind changing the base of the PR (setting the base as 8.1.x instead of develop)

metacreo commented:
Please wait. After  8.1.1 ->  8.1.2 This fix not works.

metacreo committed:
Small rework AddressController.php

Initialize address if an id exists and prevent fill form with default country second time. If this is only form request. Add param id_country for CustomerAddressForm class.

metacreo reviewed (commented):
Need review. Basically need serious rework of this functions in CustomerAddressForm, AbstractForm and AddressControllerCore. 
In order to not touch a lot of code, the only solution for now is this.

metacreo committed:
Update AddressController.php

Optimisation

metacreo reviewed (commented):
Small optimization. Checked on 8.1.1 and 8.1.2. My account address edit, add. Checkout process edit, add.

metacreo committed:
Eliminate double call of fillWith

metacreo commented:
Seems to final of PR  ;)

kpodemski commented on controllers/front/AddressController.php:
```diff
@@ -59,17 +59,25 @@ public function init()
-        $id_address = (int) Tools::getValue('id_address');
-        // Initialize address if an id exists
+
+        $params = Tools::getAllValues();
+        $id_address = isset($params['id_address']) ? (int) $params['id_address'] : 0;
```
```suggestion
        $id_address = (int) Tools::getValue('id_address', 0);
```

you could do this

metacreo replied:
I think this good, BUT we need pass array $params... 
Ok... I will test and improve this.

metacreo replied:
no we need array of form post later... in fillform when form submit..

kpodemski commented on controllers/front/AddressController.php:
```diff
@@ -59,17 +59,25 @@ public function init()
+        // Initialize address if an id exists and prevent fill form with default country second time.
         if ($id_address) {
             $this->address_form->loadAddressById($id_address);
+            // If this is only form request. Adding param id_country for CustomerAddressForm class.
```
This looks like a hack honestly, I'll have to look deeper since I'm not sure if we should have to do that there

metacreo replied:
@kpodemski no... this param needs only for correct form fill.

metacreo commented:
@kpodemski 
 this original $this->address_form->fillWith(Tools::getAllValues()); not looks as hack for you ?

metacreo commented:
@kpodemski 
function fillWith got array from loadAddressById with 2 params id_address and id_country, parameter controller added automatically when form is opened. 
When form is submitted fillWith needs full form post array.
This PR is a small rework to correct form work and prevent to fillWith call twice.
`$id_address = (int) Tools::getValue('id_address', 0);` is incomplete not array  for fillWith is needed only for loadAddressById
but if we have  `$params = Tools::getAllValues();` array, we can use it when form submited instead of
`$this->address_form->fillWith(Tools::getAllValues());`
this shortens the code and also removes the extra query
now we have two call of Tools class. with my code only one.

metacreo committed:
Update AddressController.php

Optimisation. Remove extra call of Tools class.

metacreo committed:
Update AddressController.php

metacreo committed:
Update AddressController.php

Do magic. No more need this. ;)

metacreo committed:
Update controllers/front/AddressController.php

ok. by moving fillWith afer if (Tools::isSubmit('submitAddress')) {
all seems to resolve ... so I suggest this

Co-authored-by: Krystian Podemski <kpodemski@users.noreply.github.com>

metacreo committed:
Update AddressController.php

Resolved

metacreo commented:
looks ok
```
    public function postProcess()
    {
        $this->context->smarty->assign('editing', false);

        $id_address = (int) Tools::getValue('id_address', 0);

        // Initialize address if an id exists and prevent fill form with default country second time.
        if ($id_address) {
            $this->address_form->loadAddressById($id_address);
        }

        // Submit the address, don't care if it's an edit or add
        if (Tools::isSubmit('submitAddress')) {
            // Fill the form with data
            //$this->address_form->fillWith($params);
            $this->address_form->fillWith(Tools::getAllValues());

            if (!$this->address_form->submit()) {
                $this->errors[] = $this->trans('Please fix the error below.', [], 'Shop.Notifications.Error');
            } else {
                if ($id_address) {
                    $this->success[] = $this->trans('Address successfully updated.', [], 'Shop.Notifications.Success');
                } else {
                    $this->success[] = $this->trans('Address successfully added.', [], 'Shop.Notifications.Success');
                }

                $this->should_redirect = true;
            }
        }

        // There is no id_adress, no need to continue
        if (!$id_address) {
            return;
        }
```

Hlavtox reviewed (changes requested):
@metacreo Please fill the PR form properly and add some description. Can't process any further without description. :-)

Moreover, https://github.com/PrestaShop/PrestaShop/issues/34080 is talking about some javascript issues that were not even reproduced. It should be verified and reproduced first.

metacreo commented:
@Hlavtox  Issue reproduced in #33991 See comment [1729389392](https://github.com/PrestaShop/PrestaShop/issues/33991#issuecomment-1729389392) Just QA does not have proper qualification to reproduce this bug.
Bug described in details  #34080. No need to duplicate description. I fixed everything on my servers a long time ago, and if you don’t want to, don’t accept it...  Closing PR...

Hlavtox commented:
@metacreo I think I managed to reproduce the issue, I will get back to you.

kpodemski commented:
@Hlavtox, I'm trying to do some cleaning in PRs, did you really manage to reproduce the issue? 🤔

Hlavtox commented:
@kpodemski Well, somehow - try it https://github.com/PrestaShop/PrestaShop/issues/34080#issuecomment-1823438576

metacreo commented:
[Screencast from 28.12.2023 16:30:36.webm](https://github.com/PrestaShop/PrestaShop/assets/543458/c3c65c25-5a04-43fb-aca6-52d3bd13eb9f)

metacreo committed:
Update AddressController.php

No need second 0 param in getValue, because (int).

metacreo committed:
Fix AddressController.php

Fix mistake

Hlavtox commented:
@metacreo Could you please check, if the address form uses geolocation when creating new address?

I think that by moving the fillWith call, this logic is not processed at all.

```
Tools::isCountryFromBrowserAvailable() &&
            Country::getByIso($countryIsoCode = Tools::getCountryIsoCodeFromHeader(), true)
```

metacreo commented:
> I think that by moving the fillWith call, this logic is not processed at all.

Tested. This logic not depend on `$this->address_form->fillWith(Tools::getAllValues());` in AddressController.php

Detection works ok, but you need all lang-pack to be installed for "auto" country...

metacreo commented:
@Hlavtox I  use Moz FF. 
set accept lang header via about:config for example fr-FR or en-US or ru-RU (presta reads first value of header) or all comma separated
make sure lang pack in BO installed and setting "Set default country from browser language" is ON
presta grab lang from accept lang header first locale

CustomerAddressForm.php
```
    public function fillWith(array $params = [])
    {
@dump(Tools::isCountryFromBrowserAvailable());
@dump(Tools::getCountryIsoCodeFromHeader());
```
Try move `$this->address_form->fillWith(Tools::getAllValues());` in AddressController.php
Try change browser locale.

ChillCode commented:
Hi @metacreo 

Since you're a dev, you'll understand what I expose on the following closed issue https://github.com/PrestaShop/PrestaShop/issues/34636:

Don't really now why it was closed, but it's ok, but I can solve this issue without touching that, just removing $has_currency, but that can lead to other issues, so i just unchecked browser detection from PS settings.

The issue is that we don't have country on the context because of that check so removing it we will get the same country id on each call to fillWith.

metacreo commented:
@ChillCode hello,
Please continue here #34636

metacreo commented:
I found out where this whole story came from #27187

metacreo committed:
Merge branch 'PrestaShop:develop' into patch-1
